distributed/minGPT-ddp/mingpt/main.py

```diff
@@ -0,0 +1,41 @@
+from model import GPT, GPTConfig, OptimizerConfig, create_optimizer
+from trainer import Trainer, TrainerConfig
+from char_dataset import CharDataset, DataConfig
+from torch.utils.data import random_split
+from omegaconf import DictConfig
+import hydra
+from torch.distributed import init_process_group, destroy_process_group
+
+def ddp_setup():
+    init_process_group(backend="nccl")
+
+def get_train_objs(gpt_cfg: GPTConfig, opt_cfg: OptimizerConfig, data_cfg: DataConfig):
+    dataset = CharDataset(data_cfg)
+    train_len = int(len(dataset) * data_cfg.train_split)
+    train_set, test_set = random_split(dataset, [train_len, len(dataset) - train_len])
+
+    gpt_cfg.vocab_size = dataset.vocab_size
+    gpt_cfg.block_size = dataset.block_size
+    model = GPT(gpt_cfg)
+    optimizer = create_optimizer(model, opt_cfg)
+    
+    return model, optimizer, train_set, test_set
+
+@hydra.main(version_base=None, config_path=".", config_name="gpt2_train_cfg")
```

Comment:
Another example of how this feels more like a recipe instead of an example @hudeven

Reply:
for my understanding, how do we define recipes and examples?

Reply:
I tried to explain this at the beginning of the main `README.md` but I can't say I'm super convinced that this delineation makes sense anymore. But I'd like @hudeven to make this call.

Reply:
I am hesitant to call these "recipes" as per the definition there, since these examples don't make use of torchx or other tools we have for production-first users

Reply:
@suraj813 Here are my definitions.
pytorch/examples: basic examples showcasing how to use pytorch. mainly depends on pytorch
recipes: end-to-end examples showcasing **pytorch ecosystem**(torchx, torchdata, domains, cloud setup, etc). 

Recipes are flexible on dependencies and topics, like hosting some hot applications. e.g. fine tune a stable diffusion model with checkpoint from HuggingFace, deploy a model with torchserve or pytorch mobile, distributed training in AWS with various clusters/schedulers.

For docs in pytorch/tutorials, their source code can be hosted in either repos. The basic ones go to the former and the complicated ones go to the later. What do you think?

cc: @shaojing

Reply:
In that case, I'd reckon this to be closer to a recipe because 
- it contains files talking about distributed training on an AWS cluster 
- there's potential to include pytorch ecosystem projects here too (eg: torchsnapshot as in one of the comments above, and perhaps more)

Reply:
I think it's okay to put it in examples for now. I plan to clean torchrecipes repo soon and we can move these tutorials there after that. @msaroufim what do you think?